### PR TITLE
WT-11657 add code to generate a code change info json file

### PR DESCRIPTION
**Pull request by jeremythorp**
This PR adds a new Evergreen task called 'code-change-report' that:
* Uses the results from the code coverage task ('coverage-report')
* For a patch build, obtains the diff to back to the branch from develop. For a post-merge build, uses the latest commit as the diff.
* Adds a new python script that processes the diff and the coverage data to create a new json file called 'code-change-info.json'
* Adds documentation in a README.md file.

Note: the changes to src/history/hr_conn.c are temporary and will be removed before merging. They are there to demonstrate and test what happens when C code, that is mostly covered by tests, is changed.

### CODE REVIEW

File: test/evergreen.yml [hunk 1]
```diff
@@ -3732,8 +3730,76 @@ tasks:
...
+              sed 's/$/<br>/' coverage_report/diff.txt > coverage_report/diff.html
+              # Logging for debugging
+              ls -l coverage_report
+              cat coverage_report/diff.txt
+            fi
+
+            ${python_binary|python3} test/evergreen/code_change_report/code_change_info.py -v -c coverage_report/full_coverage_report.json -g . $EXTRA_CODE_CHANGE_PARAMETERS -o coverage_report/code_change_info.json
+            
+            # Log the contents of the change info file
+            cat coverage_report/code_change_info.json
```
[the remark this answers is not in the document]
Comment (jeremythorp) [reply test/evergreen.yml:3801]: I'm inclined to keep this for now (ie in this PR) as it makes it easier to see if something goes wrong, but we have the option to remove it down the track.

File: test/evergreen/code_change_report/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      * `branches`: a list containing a dictionary per line of branch in this line of code. 
+        Note, the list will be empty of there is only one branch in the line of code. 
+        The list will either be empty or have at least two members. 
+        Each item in the list will contain the following key/value pairs:
+        * `count`: the number of times code in this branch was executed in the code coverage tests
+        * `fallthrough`: from gcovr - ignored by `code_change_info.json`
+        * `throw`: from gcovr - ignored by `code_change_info.json`
+
+Note: 
+* According to the gcovr documentation, is possible for gcov to negative code coverage count values
```
[the remark this answers is not in the document]
Comment (jeremythorp) [reply test/evergreen/code_change_report/README.md:60]: Well spotted!

File: test/evergreen/code_change_report/code_change_info.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        patch_info = get_git_info(git_working_tree_dir=git_working_tree_dir, verbose=verbose)
+    else:
+        file = open(git_diff, mode="r")
+        data = file.read()
+        diff = Diff.parse_diff(data)
+        patch_info = diff_to_change_list(diff, verbose)
+
+    report_info = create_report_info(patch_info=patch_info, coverage_data=coverage_data)
+
+    if args.outfile is not None:
```
[the remark this answers is not in the document]
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py:200]: I agree, it makes sense to make it required.

File: test/evergreen/code_change_report/code_change_info.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+
+# This function reads a gcovr json file into a dict and returns it
+def read_coverage_data(coverage_data_path: str):
+    with open(coverage_data_path) as json_file:
+        data = json.load(json_file)
+        return data
+
+
+def find_function(function_list, file_path: str, line_number: int):
```
[the remark this answers is not in the document]
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py:15]: The code was copied from a POC I built elsewhere, and in this usage this code is dead. I'll remove it.

File: test/evergreen/code_change_report/code_change_info.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        print('  Coverage data file:  {}'.format(args.coverage))
+        print('  Git root path:  {}'.format(args.git_root))
+        print('  Git diff path:  {}'.format(git_diff))
+        print('  Output file:  {}'.format(args.outfile))
+
+    git_working_tree_dir = args.git_root
+    coverage_data = read_coverage_data(args.coverage)
+
+    if git_diff is None:
+        patch_info = get_git_info(git_working_tree_dir=git_working_tree_dir, verbose=verbose)
```
[the remark this answers is not in the document]
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py:191]: That's a good idea and makes the naming more consistent.

File: test/evergreen/code_change_report/code_change_info.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajmorton) [reply test/evergreen/code_change_report/code_change_info.py]: One small type: We've added `required=True` to `--git_diff` instead of `--outfile`

**Review by ajmorton — APPROVED** [review]
LGTM with that one last change. No need for re-review 👍

File: test/evergreen/code_change_report/code_change_info.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeremythorp) [reply test/evergreen/code_change_report/code_change_info.py]: Oops - thanks for spotting that. Now fixed.

**Pull request merged**